src/main/java/io/cdap/plugin/gcp/bigquery/sqlengine/builder/BigQueryWindowsAggregationSQLBuilder.java

```diff
@@ -0,0 +1,151 @@
+
+  }
+
+  protected BigQueryWindowsAggregationSQLBuilder(WindowAggregationDefinition windowAggregationDefinition,
+                                                 String source,
+                                                 String sourceAlias,
+                                                 String rowNumColumnAlias) {
+    this.windowAggregationDefinition = windowAggregationDefinition;
+    this.source = source;
+    this.sourceAlias = sourceAlias;
+    // This is the alias used to store the row number value. Format is "rn_<uuid>"
+    this.rowNumColumnAlias = rowNumColumnAlias;
+    this.builder = new StringBuilder();
+  }
+
+  @VisibleForTesting
+  protected String getSelectedFields(Map<String, Expression> selectedFields) {
+    List<Expression> e = new ArrayList<>(selectedFields.values());
+    return getPartitionFields(e);
+  }
+
+  @Override
+  public String getQuery() {
+    String overClause = getOverClause();
+    String aggregateFields = getAggregateFields(overClause);
+    String query = SELECT + getSelectedFields(windowAggregationDefinition.getSelectExpressions()) + aggregateFields +
+      SPACE + FROM + OPEN_GROUP + SPACE + source + SPACE + CLOSE_GROUP + SPACE + AS + sourceAlias;
+    LOG.debug("Query is " + query);
+    return query;
+  }
+
+  @VisibleForTesting
+  protected String getAggregateFields(String overClause) {
+    String aggregateString = "";
+    Map<String, Expression> aggregateExpressions = windowAggregationDefinition.getAggregateExpressions();
+    for (String s : aggregateExpressions.keySet()) {
+      SQLExpression e = (SQLExpression) aggregateExpressions.get(s);
+      if (!"".equals(aggregateString)) {
+        aggregateString = aggregateString + COMMA;
+      }
+        aggregateString = aggregateString + SPACE + e.extract() + SPACE + overClause + AS + s;
+
+    }
+    return aggregateString;
+  }
+  
+  private String getOverClause() {
+    return OVER +
+      OPEN_GROUP + SPACE + PARTITION_BY + SPACE +
+      getPartitionFields(windowAggregationDefinition.getPartitionExpressions()) + SPACE
+      + getOrderByFields(windowAggregationDefinition.getOrderByExpressions()) + SPACE +
+      getWindowFrameDefinition(windowAggregationDefinition) + SPACE + CLOSE_GROUP;
+  }
+
+  private String getWindowFrameDefinition(WindowAggregationDefinition windowAggregationDefinition) {
+    String def = "";
+    if (windowAggregationDefinition.getWindowFrameType() == WindowAggregationDefinition.WindowFrameType.NONE) {
+      return def;
+    }
+    if (windowAggregationDefinition.getWindowFrameType() == WindowAggregationDefinition.WindowFrameType.RANGE) {
+      def = def + "RANGE";
+    } else if (windowAggregationDefinition.getWindowFrameType() == WindowAggregationDefinition.WindowFrameType.ROW) {
+      def = def + "ROWS";
+    }
+    def = def + SPACE + "BETWEEN";
+    if (windowAggregationDefinition.getUnboundedPreceding()) {
+      def = def + SPACE + "UNBOUNDED PRECEDING";
```

Comment:
Same here. All of these strings should be static final

Reply:
Done